railties/lib/rails/application.rb

```diff
@@ -590,5 +590,81 @@ def build_request(env)
       def build_middleware
         config.app_middleware + super
       end
+
+      class NonSymbolAccessDeprecatedHash < Hash # :nodoc:
```

Comment:
Implementation seems similar to HWA, why not inheriting from it and just override the `convert_key` method ?

Reply:
That's exactly what I tried first and decided against later, for 2 reasons:

1. HWIA stores keys as strings, but we want symbols. So it felt like I was resisting the base class.
2. Even more importantly, `convert_key` is called during both read *and* write, and there is no way to know from which it is called. That would mean we would show deprecation notices as we are interning the YAML parsed hashes and thus showing notices to people who don't even access hashes via string keys. That would be confusing.

Basically HWIA just was not flexible enough for this purpose. But I did take most of the implementation straight from it.

Hope this addresses your concern.

Reply:
I applied this change to your branch.

> 1\. HWIA stores keys as strings, but we want symbols. So it felt like I was resisting the base class.

This seems to work as expected in my tests, and passes the test suite: `Rails.application.config.my_custom_config[:foo] => {:bar=>["baz"] }`

> 2\. Even more importantly, `convert_key` is called during both read _and_ write, and there is no way to know from which it is called. That would mean we would show deprecation notices as we are interning the YAML parsed hashes and thus showing notices to people who don't even access hashes via string keys. That would be confusing.

I'm not seeing this behaviour. Can you explain how this would happen?

We can revert this to not extend from HWIA if need be, but it seems to work as expected AFAICT.

Reply:
So the `[]=` method of HWIA calls `convert_key`  here: https://github.com/rails/rails/blob/master/activesupport/lib/active_support/hash_with_indifferent_access.rb#L94 which would be called ultimately from a source that has nested hashes. If those hashes don't have string keys, we would get a deprecation notice. I guess what's protecting the code from doing that is the fact that you are calling `.to_sym` before passing `key` to `super`, so that writes always get a symbol key.

I guess that solves it more elegantly than I had it. Thanks for the rework and the simplification. 👍